raft_test.go

```diff
@@ -2395,6 +2406,275 @@ func TestRaft_ProtocolVersion_Upgrade_2_3(t *testing.T) {
+	actual := future.Error().Error()
+	if !strings.Contains(actual, expected) {
+		t.Errorf("leadership transfer should err with: %s", expected)
+	}
+}
+
+func TestRaft_LeadershipTransferLeaderRejectsClientRequests(t *testing.T) {
+	c := MakeCluster(3, t, nil)
+	defer c.Close()
+	l := c.Leader()
+	l.setLeadershipTransferInProgress(true)
+
+	// tests for API > protocol version 3 is missing here because leadership transfer
+	// is only available for protocol version >= 3
+	// TODO: is something missing here?
+	futures := []Future{
+		l.AddNonvoter(ServerID(""), ServerAddress(""), 0, 0),
+		l.AddVoter(ServerID(""), ServerAddress(""), 0, 0),
+		l.Apply([]byte("test"), 0),
+		l.Barrier(0),
+		l.DemoteVoter(ServerID(""), 0, 0),
+		l.GetConfiguration(),
+
+		// the API is tested, but here we are making sure we reject any config change.
+		l.requestConfigChange(configurationChangeRequest{}, 100*time.Millisecond),
+	}
+	futures = append(futures, l.LeadershipTransfer())
+	select {
+	case <-l.leadershipTransferCh:
+	default:
+	}
+
+	futures = append(futures, l.LeadershipTransferToServer(ServerID(""), ServerAddress("")))
+
+	for i, f := range futures {
+		if f.Error() != ErrLeadershipTransferInProgress {
+			t.Errorf("case %d: should have errored with: %s, instead of %s", i, ErrLeadershipTransferInProgress, f.Error())
+		}
+	}
+}
+
+func TestRaft_LeadershipTransferLeaderReplicationTimeout(t *testing.T) {
+	c := MakeCluster(3, t, nil)
+	defer c.Close()
+
+	l := c.Leader()
+	behind := c.GetInState(Follower)[0]
+
+	// Commit a lot of things, so that the timeout can kick in
+	for i := 0; i < 10000; i++ {
+		l.Apply([]byte(fmt.Sprintf("test%d", i)), 0)
+	}
+
+	// set ElectionTimeout really short because this is used to determine
+	// how long a transfer is allowed to take.
+	l.conf.ElectionTimeout = 1 * time.Nanosecond
+
+	future := l.LeadershipTransferToServer(behind.localID, behind.localAddr)
+	if future.Error() == nil {
+		t.Log("This test is fishing for a replication timeout, but this is not guaranteed to happen.")
+	} else {
+		expected := "leadership transfer timeout"
+		actual := future.Error().Error()
+		if !strings.Contains(actual, expected) {
+			t.Errorf("leadership transfer should err with: %s", expected)
+		}
+	}
+}
+
+func TestRaft_LeadershipTransferStopRightAway(t *testing.T) {
+	r := Raft{leaderState: leaderState{}}
+	r.setupLeaderState()
+
+	stopCh := make(chan struct{})
+	doneCh := make(chan error, 1)
+	close(stopCh)
+	r.leadershipTransfer(ServerID("a"), ServerAddress(""), stopCh, doneCh)
+	err := <-doneCh
+	if err != nil {
+		t.Errorf("leadership shouldn't have started, but instead it error with: %v", err)
+	}
+}
+
+func TestRaft_LeadershipTransferLeaderLeadershipLossDuringTransfer(t *testing.T) {
+	t.Skip("How do I test this?")
+}
+
+func TestRaft_LeadershipTransferReplicationFails(t *testing.T) {
+	t.Skip("How do I test this?")
+}
+
+func TestRaft_LeadershipTransferToUnresponsiveServer(t *testing.T) {
+	t.Skip("How do I test this?")
+}
```

Comment:
Should we just remove these if we have no way to test them?

Reply:
agreed!